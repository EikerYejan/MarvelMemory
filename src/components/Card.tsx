import React from 'react'
import styled from 'styled-components'
import { Card as CardType } from '../types'

type Props = React.HTMLAttributes<HTMLDivElement> & CardType

const Wrapper = styled.div<{ isSelected?: CardType['isSelected']; isMatched?: CardType['isMatched'] }>`
  width: 200px;
  height: 200px;
  perspective: 1000px;
  cursor: pointer;
  transform: ${({ isMatched }) => (isMatched ? 'scale(1.05)' : 'none')};
  transition: 0.1s;

  .inner {
    position: relative;
    width: 100%;
    height: 100%;
    text-align: center;
    transition: transform 0.4s;
    transform-style: preserve-3d;
    transform: ${({ isSelected }) => (isSelected ? 'rotateY(180deg)' : 'none')};
  }

  .front,
  .back {
    border-radius: 4px;
    width: 100%;
    height: 100%;
    position: absolute;
    backface-visibility: hidden;
    top: 0;
    left: 0;
    display: flex;
    justify-content: center;
    align-items: center;
  }

  .front {
    position: relative;

    img {
      width: 100%;
      height: 100%;
      object-fit: cover;
      user-select: none;
      --webkit-user-drag: none;
    }
  }

  .back {
    transform: rotateY(180deg);
    display: flex;
    justify-content: center;
    align-items: center;
    border: 2px solid ${({ theme, isMatched }) => (isMatched ? theme.colors.red : theme.colors.white)};

    &::after {
      content: '';
      position: absolute;
      display: block;
      background: rgba(0, 0, 0, 0.3);
      width: 100%;
      height: 100%;
      top: 0;
      left: 0;
      z-index: 0;
    }

    img {
      width: 100%;
      height: 100%;
      object-fit: contain;
      border-radius: inherit;
      position: absolute;
      left: 0;
      top: 0;
    }

    p {
      z-index: 1;
      color: ${({ theme }) => theme.colors.white};
    }
  }
`

const Card = ({ name, onClick, isSelected, isMatched, image }: Props) => {
  return (
    <Wrapper isMatched={isMatched} isSelected={isSelected || isMatched} tabIndex={0} onClick={onClick} role="button">
      <div className="inner">
        <div className="front">
          <img alt="marvel-logo" src="https://i.pinimg.com/originals/24/92/00/249200c431fe811110761709b303fcaf.jpg" />
        </div>
        <div className="back">
          <img src={image} alt={name} />
          <p>{name}</p>
        </div>
      </div>
    </Wrapper>
  )
}

export default Card
