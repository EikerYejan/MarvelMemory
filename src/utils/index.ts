import { DefaultTheme, StyledProps } from 'styled-components'
import { Card, Character } from '../types'

export const shuffleArray = <T extends Array<unknown>>(array: T) => {
  const shuffled = array

  for (let i = array.length - 1; i > 0; i -= 1) {
    const j = Math.floor(Math.random() * (i + 1))
    const temp = array[i]

    shuffled[i] = array[j]
    shuffled[j] = temp
  }

  return shuffled
}

export const replicateArray = <T extends Array<unknown>>(arr: T, times = 2) => {
  const arrays = Array(times).fill(0)
  const replicated = arrays.map(() => arr)

  return replicated.flat()
}

export const uuid = () => ((1 + Math.random()) * 0x10000 ?? 0).toString(16).substring(1)

export const transformCharacters = (characters: Character[]): Card[] => {
  return characters.map(({ name, thumbnail }) => ({
    id: uuid(),
    name,
    image: `${thumbnail.path}/standard_xlarge.${thumbnail.extension}`,
  }))
}

export const themeColor = (color: keyof DefaultTheme['colors']) => {
  return ({ theme }: StyledProps<DefaultTheme>) => {
    return theme.colors?.[color]
  }
}
