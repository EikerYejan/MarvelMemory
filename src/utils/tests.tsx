import React from 'react'
import { render } from '@testing-library/react'
import { ThemeProvider } from 'styled-components'
import theme from '../theme/defaultTheme'

export const renderWithTheme = (component: React.ReactElement) => {
  return render(<ThemeProvider theme={theme}>{component}</ThemeProvider>)
}
