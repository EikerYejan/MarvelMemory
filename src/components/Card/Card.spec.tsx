import React from 'react'
import { screen } from '@testing-library/react'
import Card from '.'
import { renderWithTheme } from '../../utils/tests'

const onClick = jest.fn()
const cardProps = { name: 'Test Card', image: '//cdn.com/image.png', id: '2', onClick }

describe('<Card />', () => {
  it('Should render card', () => {
    renderWithTheme(<Card {...cardProps} />)

    const container = screen.getByTestId('card')

    expect(container).toBeInTheDocument()
    expect(container).toHaveTextContent(cardProps.name)

    container.click()
    expect(onClick).toHaveBeenCalled()
  })

  it('Should render flipped card', () => {
    renderWithTheme(<Card {...cardProps} isMatched />)

    const container = screen.getByTestId('card')

    expect(container).toBeInTheDocument()
    expect(container).toMatchSnapshot()
  })
})
