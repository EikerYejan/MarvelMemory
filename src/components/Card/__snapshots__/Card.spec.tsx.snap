// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Card /> Should render flipped card 1`] = `
<div
  class="sc-bdnxRM hBUBjL"
  data-testid="card"
  role="button"
  tabindex="0"
>
  <div
    class="inner"
  >
    <div
      class="front"
    >
      <img
        alt="marvel-logo"
        src="https://i.pinimg.com/originals/24/92/00/249200c431fe811110761709b303fcaf.jpg"
      />
    </div>
    <div
      class="back"
    >
      <img
        alt="Test Card"
        src="//cdn.com/image.png"
      />
      <p>
        Test Card
      </p>
    </div>
  </div>
</div>
`;
